c++/src/kj/compat/http.h

```diff
@@ -492,6 +494,27 @@ class HttpHeaders {
   //   also add direct accessors for those headers.
 };
 
+struct HttpByteRange {
+  // Inclusive HTTP range
+
+  uint64_t start;
+  uint64_t end;
+
+  inline bool operator==(const HttpByteRange& other) const {
+    return start == other.start && end == other.end;
+  }
+  inline bool operator!=(const HttpByteRange& other) const {
+    return !(start == other.start && end == other.end);
+  }
+};
+
+kj::String KJ_STRINGIFY(HttpByteRange range);
+
+kj::Maybe<kj::Array<HttpByteRange>> tryParseHttpRangeHeader(kj::ArrayPtr<const char> value, uint64_t contentLength);
+// Parses a HTTP Range header into an array of satisfiable inclusive ranges. Returns nullptr if the
+// range specifier was invalid/unsatisfiable per https://www.rfc-editor.org/rfc/rfc9110#section-14.
+// Returns an empty array if at least one of the specifier's range specs covers the full range.
```

Comment:
It's kind of confusing for an empty array to mean "match everything" while a non-empty array means "match only what's in the array". It would be easy for the code that calls this to be buggy, interpreting an empty array as "match nothing", and yet still seem to work most of the time.

Is there an important reason to treat "match everything" specially, rather than return an array with one range that covers the whole content?

Reply:
I see. Another alternative then would be something like `OneOf<Array, Everything, Unsatisfiable, Invalid>`.

Reply:
Oh, I suppose.

Weird, intuitively I would have expected an unparseable header should result in a 400.